symengine/visitor.h

```diff
@@ -185,6 +185,217 @@ RCP<const Basic> coeff(const Basic &b, const Basic &x, const Basic &n);
 
 set_basic free_symbols(const Basic &b);
 
+class TransformVisitor : public BaseVisitor<TransformVisitor>
+{
+protected:
+    RCP<const Basic> result_;
+
+public:
+    TransformVisitor()
+    {
+    }
+
+    RCP<const Basic> apply(const Basic &x)
+    {
+        return apply(x.rcp_from_this());
+    }
+
+    RCP<const Basic> apply(const RCP<const Basic> &x)
+    {
+        x->accept(*this);
+        return result_;
+    }
+
+    void bvisit(const Basic &x)
+    {
+        result_ = x.rcp_from_this();
+    }
+
+    void bvisit(const Add &x)
+    {
+        vec_basic newargs;
+        for (const auto &a : x.get_args()) {
+            newargs.push_back(apply(a));
+        }
+        result_ = add(newargs);
+    }
+
+    void bvisit(const Mul &x)
+    {
+        vec_basic newargs;
+        for (const auto &a : x.get_args()) {
+            newargs.push_back(apply(a));
+        }
+        result_ = mul(newargs);
+    }
+
+    void bvisit(const Pow &x)
+    {
+        auto base_ = x.get_base(), exp_ = x.get_exp();
+        auto newarg1 = apply(base_), newarg2 = apply(exp_);
+        if (base_ != newarg1 or exp_ != newarg2)
+            result_ = pow(newarg1, newarg2);
+        else
+            result_ = x.rcp_from_this();
+    }
+
+    void bvisit(const OneArgFunction &x)
+    {
+        auto farg = x.get_arg();
+        auto newarg = apply(farg);
+        if (eq(*newarg, *farg))
+            result_ = x.rcp_from_this();
+        else
+            result_ = x.create(newarg);
+    }
+
+    template <class T>
+    void bvisit(const TwoArgBasic<T> &x)
+    {
+        auto farg1 = x.get_arg1(), farg2 = x.get_arg2();
+        auto newarg1 = apply(farg1), newarg2 = apply(farg2);
+        if (farg1 != newarg1 or farg2 != newarg2)
+            result_ = x.create(newarg1, newarg2);
+        else
+            result_ = x.rcp_from_this();
+    }
+
+    void bvisit(const MultiArgFunction &x)
+    {
+        auto fargs = x.get_args();
+        vec_basic newargs;
+        for (const auto &a : fargs) {
+            newargs.push_back(apply(a));
+        }
+        auto nbarg = x.create(newargs);
+        result_ = nbarg;
+    }
+};
+
+class ExpandasExp : public BaseVisitor<ExpandasExp, TransformVisitor>
```

Comment:
A new file. `rewrite.cpp`. And can you change the name to `RewriteAsExp` and the corresponding function to `rewrite_as_exp`? (SymPy has similar name)